mempool/cat/pool.go

```diff
@@ -0,0 +1,733 @@
+	if info == nil {
+		return false
+	}
+	return txmp.canAddTx(info.size)
+}
+
+func (txmp *TxPool) TryReinsertEvictedTx(txKey types.TxKey, tx types.Tx, peer uint16) error {
+	info := txmp.evictedTxs.Pop(txKey)
+	if info == nil {
+		return nil
+	}
+	txmp.logger.Debug("attempting to reinsert evicted tx", "txKey", fmt.Sprintf("%X", txKey))
+	wtx := newWrappedTx(
+		tx, txKey, txmp.height, info.gasWanted, info.priority, info.sender,
+	)
+	checkTxResp := &abci.ResponseCheckTx{
+		Code:      abci.CodeTypeOK,
+		Priority:  info.priority,
+		Sender:    info.sender,
+		GasWanted: info.gasWanted,
+	}
+	return txmp.addNewTransaction(wtx, checkTxResp)
+}
+
+// CheckTx adds the given transaction to the mempool if it fits and passes the
+// application's ABCI CheckTx method. This should be viewed as the entry method for new transactions
+// into the network. In practice this happens via an RPC endpoint
+func (txmp *TxPool) CheckTx(tx types.Tx, cb func(*abci.Response), txInfo mempool.TxInfo) error {
+	// Reject transactions in excess of the configured maximum transaction size.
+	if len(tx) > txmp.config.MaxTxBytes {
+		return mempool.ErrTxTooLarge{Max: txmp.config.MaxTxBytes, Actual: len(tx)}
+	}
+
+	// This is a new transaction that we haven't seen before. Verify it against the app and attempt
+	// to add it to the transaction pool.
+	key := tx.Key()
+	rsp, err := txmp.TryAddNewTx(tx, key, txInfo)
+	if err != nil {
+		return err
+	}
+
+	// push to the broadcast queue that a new transaction is ready
+	txmp.markToBeBroadcast(key)
+
+	// call the callback if it is set
+	if cb != nil {
+		cb(&abci.Response{Value: &abci.Response_CheckTx{CheckTx: rsp}})
+	}
+	return nil
+}
+
+func (txmp *TxPool) next() <-chan types.TxKey {
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	if len(txmp.txsToBeBroadcast) != 0 {
+		ch := make(chan types.TxKey, 1)
+		for key := range txmp.txsToBeBroadcast {
+			delete(txmp.txsToBeBroadcast, key)
+			ch <- key
+			return ch
+		}
+	}
+	return txmp.broadcastCh
+}
+
+func (txmp *TxPool) markToBeBroadcast(key types.TxKey) {
+	if !txmp.config.Broadcast {
+		return
+	}
+
+	select {
+	case txmp.broadcastCh <- key:
+	default:
+		txmp.broadcastMtx.Lock()
+		defer txmp.broadcastMtx.Unlock()
+		txmp.txsToBeBroadcast[key] = struct{}{}
+	}
+}
+
+// TryAddNewTx attempts to add a tx that has not already been seen before. It first marks it as seen
+// to avoid races with the same tx. It then call `CheckTx` so that the application can validate it.
+// If it passes `CheckTx`, the new transaction is added to the mempool solong as it has
+// sufficient priority and space else if evicted it will return an error
+func (txmp *TxPool) TryAddNewTx(tx types.Tx, key types.TxKey, txInfo mempool.TxInfo) (*abci.ResponseCheckTx, error) {
+	// First check any of the caches to see if we can conclude early. We may have already seen and processed
+	// the transaction if:
+	// - We are conencted to legacy nodes which simply flood the network
```

Comment:
[typo]

```suggestion
	// - We are connected to legacy nodes which simply flood the network
```

[question] what is a legacy node and why would it flood the network? Can you please elaborate here?

Reply:
Legacy, in the sense that it's either the v0 or v1 mempool. Flooding in the sense that the gossip protocol for these implementations is just to broadcast the transaction to all other nodes which when an entire network does it, creates this flooding effect.

Reply:
To avoid ambiguity for future readers, proposal to replace "legacy" with "nodes using the v0 or v1 mempool"